components/style/properties/properties.mako.rs

```diff
@@ -2034,6 +2020,181 @@ pub fn get_writing_mode(inheritedbox_style: &style_structs::InheritedBox) -> Wri
+    pub fn mutate(&mut self) -> &mut T {
+        if let StyleStructRef::Borrowed(v) = *self {
+            *self = StyleStructRef::Owned((**v).clone());
+        }
+
+        match *self {
+            StyleStructRef::Owned(ref mut v) => v,
+            StyleStructRef::Borrowed(..) => unreachable!(),
+        }
+    }
+
+    /// Returns an `Arc` to the internal struct, constructing one if
+    /// appropriate.
+    pub fn build(self) -> Arc<T> {
+        match self {
+            StyleStructRef::Owned(v) => Arc::new(v),
+            StyleStructRef::Borrowed(v) => v.clone(),
+        }
+    }
+}
+
+impl<'a, T: 'a> Deref for StyleStructRef<'a, T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        match *self {
+            StyleStructRef::Owned(ref v) => v,
+            StyleStructRef::Borrowed(v) => &**v,
+        }
+    }
+}
+
+/// A type used to compute a struct with minimal overhead.
+///
+/// This allows holding references to the parent/default computed values without
+/// actually cloning them, until we either build the style, or mutate the
+/// inherited value.
+pub struct StyleBuilder<'a> {
+    custom_properties: Option<Arc<::custom_properties::ComputedValuesMap>>,
+    /// The writing mode flags.
+    ///
+    /// TODO(emilio): Make private.
+    pub writing_mode: WritingMode,
+    /// The font size of the root element.
+    pub root_font_size: Au,
+    /// The keyword behind the current font-size property, if any.
+    pub font_size_keyword: Option<(longhands::font_size::KeywordSize, f32)>,
+    % for style_struct in data.active_style_structs():
+        ${style_struct.ident}: StyleStructRef<'a, style_structs::${style_struct.name}>,
+    % endfor
+}
+
+impl<'a> StyleBuilder<'a> {
+    /// Trivially construct a `StyleBuilder`.
+    pub fn new(
+        custom_properties: Option<Arc<::custom_properties::ComputedValuesMap>>,
+        writing_mode: WritingMode,
+        root_font_size: Au,
+        font_size_keyword: Option<(longhands::font_size::KeywordSize, f32)>,
+        % for style_struct in data.active_style_structs():
+            ${style_struct.ident}: &'a Arc<style_structs::${style_struct.name}>,
+        % endfor
+    ) -> Self {
+        StyleBuilder {
+            custom_properties: custom_properties,
+            writing_mode: writing_mode,
+            root_font_size: root_font_size,
+            font_size_keyword: font_size_keyword,
+        % for style_struct in data.active_style_structs():
+            ${style_struct.ident}: StyleStructRef::Borrowed(${style_struct.ident}),
+        % endfor
+        }
+    }
+
+    /// Creates a StyleBuilder holding only references to the structs of `s`.
+    pub fn for_default_style(s: &'a ComputedValues) -> Self {
+        Self::for_derived_style(s)
+    }
+
+    /// Creates a StyleBuilder holding only references to the structs of `s`, in
+    /// order to create a derived style.
+    pub fn for_derived_style(s: &'a ComputedValues) -> Self {
```

Comment:
Yeah, it's just because I found useful to do the discrimination between "we're using the default style" (media-queries, viewport rule), and "we're deriving from this style" (keyframes), happy to unify if needed though.